docs/proposals/workflow.md

```diff
@@ -0,0 +1,255 @@
+	// Value must be non-negative integer. The value zero indicates delete immediately.
+	// If this value is nil, the default grace period will be used instead.
+	// Set this value longer than the expected cleanup time for your workflow.
+    // If downstream resources (job, pod, etc.) define their TerminationGracePeriodSeconds
+    // the biggest is taken.
+	TerminationGracePeriodSeconds *int64 `json:"terminationGracePeriodSeconds,omitempty"`
+
+	// Steps contains the vertices of the workflow graph.
+	Steps []WorkflowStep `json:"steps,omitempty"`
+}
+```
+
+* `spec.steps`: is an array of `WorkflowStep`s.
+* `spec.terminationGracePeriodSeconds`: is the terminationGracePeriodSeconds.
+
+### `WorkflowStep`<sup>1</sup>
+
+The `WorkflowStep` resource acts as a [union](https://en.wikipedia.org/wiki/Tagged_union) of `JobSpec` and `ObjectReference`.
+
+```go
+// WorkflowStep contains necessary information to identifiy the node of the workflow graph
+type WorkflowStep struct {
+    // Id is the identifier of the current step
+    Id string  `json:"id,omitempty"`
+
+    // Spec contains the job specificaton that should be run in this Workflow.
+	// Only one between External and Spec can be set.
+	Spec JobSpec `json:"jobSpec,omitempty"`
+
+    // Dependecies represent dependecies of the current workflow step
+    Dependencies ObjectDependencies `json:"dependencies,omitempty"`
+
+	// External contains a reference to another schedulable resource.
+	// Only one between ExternalRef and Spec can be set.
+	ExternalRef api.ObjectReference `json:"externalRef,omitempty"`
+}
+```
+
+* `workflowStep.id` is a string to identify the current `Workflow`. The `workfowStep.id` is injected
+as a label in `metadata.annotations` in the `Job` created in the current step.
+* `workflowStep.jobSpec` contains the specification of the job to be executed.
+* `workflowStep.externalRef` contains a reference to external resources (for example another `Workflow`).
+
+```go
+type ObjectDependencies struct {
+    ...
+    ExistenceDependencies []ObjectReference `json:"existenceDependencies,omitempty"`
+    ControllerRef *ObjectReference `json:"controllerRef,omitempty"`
+    ...
+}
+```
+
+* `dependencies.controllerRef`: will contain the policy to trigger current `WorkflowStep`
+
+For `Workflow` basic scenario 2 controllers should be implemented: `AllDependenciesRunToCompletion`,
+`AtLeastOneDependencyRunToCompletion`. This approach permits implementation of other kinds of triggering
+policies, like for example data availability or other external event.
+
+
+
+### `WorkflowStatus`
+
+```go
+// WorkflowStatus contains the current status of the Workflow
+type WorkflowStatus struct {
+	Statuses []WorkflowStepStatus `json:statuses`
+}
+
+// WorkflowStepStatus contains the status of a WorkflowStep
+type WorkflowStepStatus struct {
+	// Job contains the status of Job for a WorkflowStep
+	JobStatus Job `json:"jobsStatus,omitempty"`
```

Comment:
The Workflow controller is going to create a Job object from the template in the WorkflowStep.  And that Job will have status.  So, I don't know that there is any value in copying it here.

What might be valuable is to have the status contain the name of (or an objectReference to) the Job object that it created, so the user can check there for status.